Fix queued script not updating current attribute when queuing

pnbruckner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
When a script configured in queued mode first queued a run it was not updating its `current` attribute (which is the current number of active runs.) This change fixes that.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pnbruckner committed:
Fix queued script not updating current attribute when queuing

frenck commented:
Can we add/maintain the test coverage for this?

pnbruckner commented:
I did update the test to cover these changes. If you look at the details of the coverage report, it's complaining about a block of code that didn't change, but got moved. So not sure why it's complaining now when it didn't before.

pnbruckner commented:
I can try adding a test that covers that small block of code.

pnbruckner committed:
Add test of cancelling queued runs

pnbruckner commented:
Now that I think about it some more, that block of code is very difficult to test. It needs to be there to handle the case where the loop is being shutdown, but it can't be reached by cancelling `Script.async_run()`, because that block is shielded.

Not sure what should be done about this, or if anything reasonably can. Like I said, the coverage report never complained about it before.

And, by the way, the new test is still valid, so I guess it doesn't hurt to leave it in.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -539,23 +539,32 @@ class _QueuedScriptRun(_ScriptRun):
+        else:
+            # No: We can just go ahead and grab the lock which will be immediately
+            # successful.
+            await queue_lock.acquire()
```
This can hit a race condition where your `await` task is queued after a current task that is acquiring it.

pnbruckner replied:
Ah, good point. Mainly I'm just trying to avoid calling self._changed() if super().async_run() is going to be called pretty much next so there aren't two state changes one right after the other.

I'll find a better way to do this. So far as I know, the bug I'm trying to fix only affects the frontend; the script runs as it's supposed to. So it's not super urgent.

pnbruckner replied:
I couldn't find a more optimal way to do it, so I just did a call to change listener immediately after updating runs list instead.

pnbruckner replied:
Well, darn, I don't like this! Now there are two `state_changed` events every time the script is run.

I'm going to try a different tack, and that's to move the update of the `last_triggered` attribute out of `_ScriptRun.async_run()` and put it in `Script.async_run()`, together with adding the new run to the list, and before calling the change listener. Then I can also remove that initial call to the change listener in `_ScriptRun.async_run()`. That should get it back to one `state_changed` event when the script starts running, _and_ making sure that ultimately the `current` attribute is updated correctly.

The main difference will be that for queued runs, `last_triggered` will be updated when the run is queued, not when it is run. But I think that makes more sense anyway.

pnbruckner replied:
Ok, that feels much cleaner and like the way it should have been in the first place. 😃

pnbruckner committed:
Update per review

Revert script helper change and instead call change listener immediately after new run added

pnbruckner committed:
Update per review attempt #2

Move update of last_triggered from _ScriptRun to Script

balloob reviewed (approved):
Nice!
